convert nxp s32 qspi to zephyr native driver

manuargue opened the pull request:
Convert NXP S32 QSPI controller and flash NOR driver to native drivers. Existing functionalities are retained and some enhancements are done to allow more configuration options. Changes are large, so is better perhaps to review it as a new driver.

```
INFO    - 2764/2778 mr_canhubk3               tests/drivers/flash/common/drivers.flash.common.mr_canhubk3_sfdp_runtime PASSED (device: AB0O4EZU, 11.070s)
INFO    - 2765/2778 mr_canhubk3               samples/subsys/settings/sample.subsys.settings     PASSED (device: AB0O4EZU, 9.580s)
INFO    - 2766/2778 mr_canhubk3               tests/subsys/fs/littlefs/filesystem.littlefs.default PASSED (device: AB0O4EZU, 19.467s)
INFO    - 2767/2778 mr_canhubk3               tests/subsys/storage/flash_map/storage.flash_map.mbedtls PASSED (device: AB0O4EZU, 47.489s)
INFO    - 2768/2778 mr_canhubk3               tests/subsys/logging/log_backend_fs/logging.backend.fs.manualmounted PASSED (device: AB0O4EZU, 18.954s)
INFO    - 2769/2778 mr_canhubk3               tests/subsys/fs/littlefs/filesystem.littlefs.custom PASSED (device: AB0O4EZU, 56.370s)
INFO    - 2770/2778 mr_canhubk3               tests/subsys/settings/functional/fcb/settings.functional.fcb PASSED (device: AB0O4EZU, 12.231s)
INFO    - 2771/2778 mr_canhubk3               tests/subsys/settings/functional/file/settings.file PASSED (device: AB0O4EZU, 14.187s)
INFO    - 2772/2778 mr_canhubk3               tests/drivers/flash/common/drivers.flash.common.default PASSED (device: AB0O4EZU, 9.184s)
INFO    - 2773/2778 mr_canhubk3               tests/subsys/settings/functional/fcb/settings.functional.fcb.chosen PASSED (device: AB0O4EZU, 13.390s)
INFO    - 2774/2778 mr_canhubk3               tests/subsys/settings/fcb/settings.fcb.raw         PASSED (device: AB0O4EZU, 34.381s)
INFO    - 2775/2778 mr_canhubk3               tests/subsys/storage/flash_map/storage.flash_map   PASSED (device: AB0O4EZU, 44.564s)
INFO    - 2776/2778 mr_canhubk3               tests/subsys/fs/fs_api/filesystem.api              PASSED (device: AB0O4EZU, 7.591s)
INFO    - 2777/2778 mr_canhubk3               tests/subsys/storage/flash_map/storage.flash_map.psa PASSED (device: AB0O4EZU, 45.087s)
INFO    - 2778/2778 mr_canhubk3               tests/subsys/settings/file/settings.file.raw       PASSED (device: AB0O4EZU, 9.577s)
```

manuargue-nxp committed:
drivers: memc: nxp_qspi: convert to native driver

Convert NXP QSPI memory controller driver to a Zephyr native driver.
Existing functionalities are retained and enhanced to allow more
configuration options, including multiple port sides.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

congnguyenhuu commented on drivers/memc/memc_nxp_s32_qspi.c:
```diff
@@ -1,197 +1,1707 @@
+#endif /* CONFIG_MEMC_NXP_S32_QSPI_HAS_DLLCR_REFCNTR */
+#if defined(CONFIG_MEMC_NXP_S32_QSPI_HAS_DLLCR_DLLRES)
+	reg_val &= ~QSPI_DLLCR_DLL_DLLRES_MASK;
+	reg_val |= QSPI_DLLCR_DLL_DLLRES(controller->ports[port].dll.resolution);
```
these macros `QSPI_DLLCR_DLL_DLLRES_MASK` and `QSPI_DLLCR_DLL_DLLRES` are not defined

congnguyenhuu commented on drivers/memc/memc_nxp_s32_qspi.c:
```diff
@@ -1,197 +1,1707 @@
+			.all_masters = (bool)DT_INST_PROP(n, ahb_buffers_all_masters),             \
+		},                                                                                 \
+		IF_ENABLED(CONFIG_MEMC_NXP_S32_QSPI_HAS_SFACR_CAS ||                               \
+			   CONFIG_MEMC_NXP_S32_QSPI_HAS_SFACR_BYTE_SWAP, (                         \
```
you can not use `IF_ENABLED` with a condition that contains the operator. please update.

manuargue-nxp committed:
drivers: flash: nxp_qspi_nor: convert to native driver

Convert NXP QSPI NOR flash driver to a Zephyr native driver,
retaining existing functionalities.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
drivers: flash: nxp_qspi_nor: implement reset op

Implement basic flash memory reset and enable it through the extended
operations API.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
drivers: flash: nxp_qspi_nor: remove readback verification

The readback verification on program and erase have no practical use and
was meant to be enabled only for testing, but the flash tests are
already reading back the memory after programming or erasing, so makes
no sense to keep it.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue commented:
- addressed all comments above
- split commits for updates in flash driver
- made the vlut for the transfer config of a fixed size to check for boundaries

manuargue commented:
- added some parameter validations & assertions
- corrected port base address function for ports > A1

manuargue commented:
- correct mask for `QSPI_WORD0` that was missed in previous push

Dat-NguyenDuy reviewed (changes requested):
Some nits but would like to see addressed

manuargue commented on drivers/memc/memc_nxp_s32_qspi.c:
```diff
@@ -1,197 +1,1674 @@
+
+	/* Software reset AHB domain and Serial Flash domain at the same time */
+	REG_WRITE(QSPI_MCR, REG_READ(QSPI_MCR) | QSPI_MCR_SWRSTHD(1U) | QSPI_MCR_SWRSTSD(1U));
+	delay = QSPI_RESET_DELAY_CYCLES;
```
I'm using the same criteria as the HAL, in fact this is inherited from it. In the RM is specified in cycles

manuargue commented on drivers/flash/flash_nxp_s32_qspi_nor.c:
```diff
@@ -615,492 +494,818 @@ static int nxp_s32_qspi_write(const struct device *dev, off_t offset, const void
+		.alignment = config->mem_alignment,
+	};
+
+	qspi_release(dev);
```
yes, fixed

Dat-NguyenDuy replied:
Have you forgot to re-push, i see it has not fixed

manuargue replied:
yes my bad, pushed

manuargue commented:
> Is this intended?

not intended, fixed

manuargue commented:
re-opened

manuargue commented:
re-opened
